File: node/cubrid/03_transactions.js
// 03_transactions.js — Transaction management with @cubrid/client.
//
// Demonstrates:
// - Automatic commit on success with db.transaction()
// - Automatic rollback on error with db.transaction()
// - Multiple statements within a single transaction

import { createClient } from "@cubrid/client";

const DB_CONFIG = {
  host: "localhost",
  port: 33000,
  database: "testdb",
  user: "dba",
  password: "",
};

async function setup(db) {
  await db.query("DROP TABLE IF EXISTS cookbook_accounts");
  await db.query(`
    CREATE TABLE cookbook_accounts (
      id      INT AUTO_INCREMENT PRIMARY KEY,
      name    VARCHAR(100) NOT NULL,
      balance DOUBLE DEFAULT 0.0
    )
  `);
  await db.query(
    "INSERT INTO cookbook_accounts (name, balance) VALUES (?, ?)",
    ["Alice", 1000.0],
  );
  await db.query(
    "INSERT INTO cookbook_accounts (name, balance) VALUES (?, ?)",
    ["Bob", 500.0],
  );
}

async function showBalances(db, label = "") {
  const rows = await db.query(
    "SELECT name, balance FROM cookbook_accounts ORDER BY id",
  );
  const tag = label ? ` (${label})` : "";
  const balances = rows
    .map((r) => `${r.name}=$${Number(r.balance).toFixed(2)}`)
    .join(", ");
  console.log(`  Balances${tag}: ${balances}`);
}

async function commitExample(db) {
  console.log("\n=== Commit Example ===");
  await showBalances(db, "before");

  const amount = 200.0;

  // transaction() auto-commits on success, auto-rolls back on error
  await db.transaction(async (tx) => {
    await tx.query(
      "UPDATE cookbook_accounts SET balance = balance - ? WHERE name = ?",
      [amount, "Alice"],
    );
    await tx.query(
      "UPDATE cookbook_accounts SET balance = balance + ? WHERE name = ?",
      [amount, "Bob"],
    );
  });

  console.log(`  ✓ Transferred $${amount.toFixed(2)} from Alice to Bob`);
  await showBalances(db, "after commit");
}

async function rollbackExample(db) {
  console.log("\n=== Rollback Example ===");
  await showBalances(db, "before");

  try {
    await db.transaction(async (tx) => {
      await tx.query(
        "UPDATE cookbook_accounts SET balance = 0 WHERE name = 'Alice'",
      );
      console.log("  Made Alice's balance = 0 (within transaction)");

      // Throwing inside the callback triggers automatic rollback
      throw new Error("Something went wrong — rolling back!");
    });
  } catch (_err) {
    console.log("  ✓ Transaction rolled back automatically on error");
  }

  await showBalances(db, "after rollback");
}

async function multiStatementTransaction(db) {
  console.log("\n=== Multi-Statement Transaction ===");
  await showBalances(db, "before");

  // All three updates commit together or not at all
  await db.transaction(async (tx) => {
    await tx.query(
      "UPDATE cookbook_accounts SET balance = balance + 100 WHERE name = 'Alice'",
    );
    await tx.query(
      "UPDATE cookbook_accounts SET balance = balance + 50 WHERE name = 'Bob'",
    );
    console.log("  Applying bonuses: Alice +$100, Bob +$50");
  });

  await showBalances(db, "after bonuses");
}

async function cleanup(db) {
  await db.query("DROP TABLE IF EXISTS cookbook_accounts");
  console.log("\n✓ Cleaned up");
}

async function main() {
  const db = createClient(DB_CONFIG);
  try {
    await setup(db);
    await commitExample(db);
    await rollbackExample(db);
    await multiStatementTransaction(db);
  } finally {
    await cleanup(db);
    await db.close();
  }
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: python/pycubrid/03_transactions.py
"""03_transactions.py — Transaction management with pycubrid.

Demonstrates:
- Manual commit/rollback
- Autocommit mode
- Savepoints (SAVEPOINT, ROLLBACK TO SAVEPOINT)
- Transaction isolation
"""

from __future__ import annotations

import pycubrid

DB_CONFIG = {
    "host": "localhost",
    "port": 33000,
    "database": "testdb",
    "user": "dba",
    "password": "",
}


def get_connection() -> pycubrid.Connection:
    return pycubrid.connect(**DB_CONFIG)


def setup(conn: pycubrid.Connection) -> None:
    cursor = conn.cursor()
    cursor.execute("DROP TABLE IF EXISTS cookbook_accounts")
    cursor.execute("""
        CREATE TABLE cookbook_accounts (
            id      INT AUTO_INCREMENT PRIMARY KEY,
            name    VARCHAR(100) NOT NULL,
            balance DOUBLE DEFAULT 0.0
        )
    """)
    cursor.execute("INSERT INTO cookbook_accounts (name, balance) VALUES ('Alice', 1000.0)")
    cursor.execute("INSERT INTO cookbook_accounts (name, balance) VALUES ('Bob', 500.0)")
    conn.commit()
    cursor.close()


def show_balances(conn: pycubrid.Connection, label: str = "") -> None:
    cursor = conn.cursor()
    cursor.execute("SELECT name, balance FROM cookbook_accounts ORDER BY id")
    rows = cursor.fetchall()
    tag = f" ({label})" if label else ""
    print(f"  Balances{tag}: {', '.join(f'{r[0]}=${r[1]:.2f}' for r in rows)}")
    cursor.close()


def commit_example(conn: pycubrid.Connection) -> None:
    """Transfer money with explicit commit."""
    print("\n=== Commit Example ===")
    show_balances(conn, "before")

    cursor = conn.cursor()
    amount = 200.0

    # Debit Alice
    cursor.execute(
        "UPDATE cookbook_accounts SET balance = balance - ? WHERE name = ?",
        (amount, "Alice"),
    )
    # Credit Bob
    cursor.execute(
        "UPDATE cookbook_accounts SET balance = balance + ? WHERE name = ?",
        (amount, "Bob"),
    )

    # Both changes are visible within this transaction
    conn.commit()
    print(f"  ✓ Transferred ${amount:.2f} from Alice to Bob")

    show_balances(conn, "after commit")
    cursor.close()


def rollback_example(conn: pycubrid.Connection) -> None:
    """Demonstrate rollback — changes are discarded."""
    print("\n=== Rollback Example ===")
    show_balances(conn, "before")

    cursor = conn.cursor()

    # Make a bad update
    cursor.execute("UPDATE cookbook_accounts SET balance = 0 WHERE name = 'Alice'")
    print("  Made Alice's balance = 0 (not committed)")

    # Oops — rollback!
    conn.rollback()
    print("  ✓ Rolled back — Alice's balance restored")

    show_balances(conn, "after rollback")
    cursor.close()


def autocommit_example() -> None:
    """Autocommit mode — each statement commits immediately."""
    print("\n=== Autocommit Example ===")
    conn = get_connection()
    conn.autocommit = True

    cursor = conn.cursor()
    cursor.execute("UPDATE cookbook_accounts SET balance = balance + 50 WHERE name = 'Alice'")
    print("  ✓ Updated Alice +$50 (auto-committed immediately)")

    # No need to call conn.commit()
    show_balances(conn, "autocommit")

    conn.autocommit = False  # Back to manual mode
    cursor.close()
    conn.close()


def savepoint_example(conn: pycubrid.Connection) -> None:
    """Savepoints — partial rollback within a transaction."""
    print("\n=== Savepoint Example ===")
    show_balances(conn, "before")

    cursor = conn.cursor()

    # Step 1: Give Alice a bonus
    cursor.execute("UPDATE cookbook_accounts SET balance = balance + 100 WHERE name = 'Alice'")
    print("  Step 1: Alice +$100")

    # Create savepoint after step 1
    cursor.execute("SAVEPOINT after_alice_bonus")
    print("  ✓ Created SAVEPOINT 'after_alice_bonus'")

    # Step 2: Give Bob a bonus (we'll undo this)
    cursor.execute("UPDATE cookbook_accounts SET balance = balance + 999 WHERE name = 'Bob'")
    print("  Step 2: Bob +$999 (will be rolled back)")

    # Rollback to savepoint — undoes step 2 but keeps step 1
    cursor.execute("ROLLBACK TO SAVEPOINT after_alice_bonus")
    print("  ✓ Rolled back to savepoint — Bob's $999 undone, Alice's $100 kept")

    conn.commit()
    show_balances(conn, "after savepoint rollback + commit")
    cursor.close()


def cleanup(conn: pycubrid.Connection) -> None:
    cursor = conn.cursor()
    cursor.execute("DROP TABLE IF EXISTS cookbook_accounts")
    conn.commit()
    cursor.close()
    print("\n✓ Cleaned up")


if __name__ == "__main__":
    conn = get_connection()

    try:
        setup(conn)
        commit_example(conn)
        rollback_example(conn)
        autocommit_example()
        savepoint_example(conn)
    finally:
        cleanup(conn)
        conn.close()
